dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/EnrollmentInExistingValidationHook.java

```diff
@@ -104,107 +76,67 @@ public void validateEnrollment( ValidationErrorReporter reporter, Enrollment enr
         validateTeiNotEnrolledAlready( reporter, enrollment, program );
     }
 
-    protected void validateTeiNotEnrolledAlready( ValidationErrorReporter reporter,
+    private void validateTeiNotEnrolledAlready( ValidationErrorReporter reporter,
         Enrollment enrollment, Program program )
     {
-        User user = reporter.getValidationContext().getBundle().getUser();
-
-        checkNotNull( user, USER_CANT_BE_NULL );
-        checkNotNull( program, PROGRAM_CANT_BE_NULL );
         checkNotNull( enrollment.getTrackedEntity(), TRACKED_ENTITY_INSTANCE_CANT_BE_NULL );
 
         TrackedEntityInstance tei = getTrackedEntityInstance( reporter, enrollment.getTrackedEntity() );
 
-        // TODO: Create a dedicated sql query....?
-        Set<Enrollment> activeAndCompleted = getAllEnrollments( reporter, program, tei.getUid() )
+        Set<Enrollment> payloadEnrollment = reporter.getValidationContext().getBundle().getEnrollments()
+            .stream().filter( Objects::nonNull )
+            .filter( pi -> pi.getProgram().equals( program.getUid() ) )
+            .filter( pi -> pi.getTrackedEntity().equals( tei.getUid() )
+                && !pi.getUid().equals( enrollment.getEnrollment() ) )
+            .filter( pi -> EnrollmentStatus.ACTIVE == pi.getStatus() || EnrollmentStatus.COMPLETED == pi.getStatus() )
+            .collect( Collectors.toSet() );
+
+        Set<Enrollment> dbEnrollment = reporter.getValidationContext().getBundle().getPreheat()
+            .getTrackedEntityToProgramInstanceMap().getOrDefault( enrollment.getTrackedEntity(), new ArrayList<>() )
             .stream()
-            .filter( e -> !e.getEnrollment().equals( enrollment.getEnrollment() )
-                && (EnrollmentStatus.ACTIVE == e.getStatus() || EnrollmentStatus.COMPLETED == e.getStatus()) )
+            .filter( Objects::nonNull )
+            .filter( pi -> pi.getProgram().getUid().equals( program.getUid() )
+                && !pi.getUid().equals( enrollment.getEnrollment() ) )
+            .filter( pi -> ProgramStatus.ACTIVE == pi.getStatus() || ProgramStatus.COMPLETED == pi.getStatus() )
+            .collect( Collectors.toSet() ).stream().map( this::getEnrollmentFromProgramInstance )
```

Comment:
Maybe using distinct is less verbose

```suggestion
            .distinct().map( this::getEnrollmentFromProgramInstance )
```